### PR TITLE
[CAS] Teach swift-frontend to replay result from cache

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
Teach swift-frontend to replay cache results if there is a cache hit.

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

File: lib/Frontend/CompileJobCacheKey.cpp [hunk 1]
```diff
@@ -48,6 +48,13 @@ llvm::Expected<llvm::cas::ObjectRef> swift::createCompileJobBaseCacheKey(
       SkipNext = true;
       continue;
     }
+    // FIXME: Use a heuristic to remove all the flags that affect output paths.
+    // Those should not affect compile cache key.
+    if (Arg.startswith("-emit-")) {
```
Comment (artemcm) [inline lib/Frontend/CompileJobCacheKey.cpp:53]: 😬

File: lib/Frontend/CachingUtils.cpp [hunk 2]
```diff
@@ -133,6 +133,134 @@ createSwiftCachingOutputBackend(
...
+      return None;
+    }
+    auto OutputKeyID = CAS.getID(*OutputKey);
+    auto Lookup = Cache.get(OutputKeyID);
+    if (!Lookup) {
+      Diag.diagnose(SourceLoc(), diag::error_cas, toString(Lookup.takeError()));
+      return None;
+    }
+    if (!*Lookup) {
+      Diag.diagnose(SourceLoc(), diag::output_cache_miss, OutputPath,
```
Comment (artemcm) [inline lib/Frontend/CachingUtils.cpp:170]: Going forward, should these remarks be being a flag of some sort?

**Review by artemcm — APPROVED** [review]
This is very cool.

File: lib/Frontend/CachingUtils.cpp (hunk 2 above)
Comment (cachemeifyoucan) [reply lib/Frontend/CachingUtils.cpp:170]: Yeah, we should guard them but at some point I will reconfigure all the flags. Currently I am thinking that:
* add `-compile-cache` driver option to turn on replay and output storing.
* `-enable-cas` will just be a frontend option to add a cas instance for various cas tasks, like cas based dependency scanning, cas fs, cas object file format, etc.

**Pull request merged**